src/Bicep.Core.RegistryClient/BicepRegistryBlobClient.cs

```diff
@@ -0,0 +1,97 @@
+            Argument.AssertNotNull(credential, nameof(credential));
+            Argument.AssertNotNull(options, nameof(options));
+
+            _endpoint = endpoint;
+            _registryName = endpoint.Host.Split('.')[0];
+            _repositoryName = repositoryName;
+            _clientDiagnostics = new ClientDiagnostics(options);
+
+            _acrAuthPipeline = HttpPipelineBuilder.Build(options);
+            _acrAuthClient = new AuthenticationRestClient(_clientDiagnostics, _acrAuthPipeline, endpoint.AbsoluteUri);
+
+            _pipeline = HttpPipelineBuilder.Build(options, new ContainerRegistryChallengeAuthenticationPolicy(credential, options.AuthenticationScope, _acrAuthClient));
+            _restClient = new ContainerRegistryRestClient(_clientDiagnostics, _pipeline, _endpoint.AbsoluteUri);
+            _blobRestClient = new ContainerRegistryBlobRestClient(_clientDiagnostics, _pipeline, _endpoint.AbsoluteUri);
+        }
+
+        public async Task<Response<UploadBlobResult>> UploadBlobAsync(Stream stream, CancellationToken cancellationToken = default)
+        {
+            string digest = DigestHelper.ComputeDigest(DigestHelper.AlgorithmIdentifierSha256, stream);
+
+            ResponseWithHeaders<ContainerRegistryBlobStartUploadHeaders> startUploadResult =
+                await _blobRestClient.StartUploadAsync(_repositoryName, cancellationToken).ConfigureAwait(false);
+
+            stream.Position = 0;
+            ResponseWithHeaders<ContainerRegistryBlobUploadChunkHeaders> uploadChunkResult =
+                await _blobRestClient.UploadChunkAsync(startUploadResult.Headers.Location, stream, cancellationToken).ConfigureAwait(false);
+
+            ResponseWithHeaders<ContainerRegistryBlobCompleteUploadHeaders> completeUploadResult =
+                await _blobRestClient.CompleteUploadAsync(digest, uploadChunkResult.Headers.Location, null, cancellationToken).ConfigureAwait(false);
+
+            return Response.FromValue(new UploadBlobResult(), completeUploadResult.GetRawResponse());
+        }
+
+        public async Task<Response<UploadManifestResult>> UploadManifestAsync(Stream stream, UploadManifestOptions options = default, CancellationToken cancellationToken = default)
+        {
+            options ??= new UploadManifestOptions();
+
+            string reference = options.Tag ?? DigestHelper.ComputeDigest(DigestHelper.AlgorithmIdentifierSha256, stream);
+            stream.Position = 0;
+            ResponseWithHeaders<ContainerRegistryCreateManifestHeaders> response = await _restClient.CreateManifestAsync(_repositoryName, reference, options.MediaType, stream, cancellationToken).ConfigureAwait(false);
+
+            return Response.FromValue(new UploadManifestResult(), response.GetRawResponse());
+        }
+
+        public async Task<Response<DownloadManifestResult>> DownloadManifestAsync(string reference, DownloadManifestOptions options = default, CancellationToken cancellationToken = default)
```

Comment:
>reference [](http://example.com/codeflow?start=89&length=9)

@annelo-msft I made a change here to allow tags and digests. I also needed full access to the manifest content to read media types, annotations, etc.), so I ended up removing the `GetManifestProperties()` call and the `ArtifactFiles` properties  from `DownloadManifestResult`.

The end result ended up removing one REST call from this method and one REST call on my end to get the digest for a particular tag.

Reply:
I think there may be a case to have a simpler version of `DownloadManifestAsync` for tools that need to deeply inspect the manifest and maybe a different version of it for callers that aren't interested in those details?